### PR TITLE
break up `create_clean_reported_cases()`

### LINKED ISSUE #346: reported_cases_opts()
Currently, all inputs except for `reported_cases` are managed via a helper function (i.e `delay_opts()`) that enables specifying options. It would make sense to standardized `reported_cases` to be in line with this approach. This would also make sense as currently there are several data processing steps (i.e to deal with missing dates) that occur inside package code that are not well surfaced to the user. Putting these in a `reported_cases_opts()` function would help resolve this and bring the user closer to the data being used for model fitting which is generally a good idea.

### PR DESCRIPTION
**Pull request by sbfnk**
## Description

This PR closes #346 by breaking up the `create_clean_reported_cases()` function into its individual steps that are now separate functions that can be applied at preprocessing stage. After a deprecation cycle this function will be removed.

Doing so will facilitate addressing #640 via the new `add_horizon()` function that allows users to specify a forecast interval (internal for now, could later be exposed or accessed via `forecast_opts()` or the like when addressing #867).

It will also facilitate addressing #451 as data passed to functions can be used for plotting as preprocessing happens before calling `estimate_infections()` and the like.

## Initial submission checklist

- [X] My PR is based on a package issue and I have explicitly linked it.
- [X] I have tested my changes locally (using `devtools::test()` and `devtools::check()`).
- [X] I have added or updated unit tests where necessary.
- [X] I have updated the documentation if required and rebuilt docs if yes (using `devtools::document()`).
- [X] I have followed the established coding standards (and checked using `lintr::lint_package()`).
- [X] I have added a news item linked to this PR.

## After the initial Pull Request 

- [x] I have reviewed Checks for this PR and addressed any issues as far as I am able.

### CODE REVIEW

File: R/preprocessing.R [hunk 1]
```diff
@@ -177,3 +177,154 @@ default_fill_missing_obs <- function(data, obs, obs_column) {
...
+##'   are meaningful by using a threshold number of cases based on the 7-day
+##'   average. If the average is above this threshold at the time of a zero
+##'   observation count then the zero is replaced with a missing (`NA`) count
+##'   and thus ignored in the likelihood.
+##'
+##' @inheritParams estimate_infections
+##' @inheritParams fill_missing
+##' @importFrom data.table setDT frollsum
+##' @return A data.table with the zero threshold applied.
+##' @author Sebastian Funk
```
Comment (seabbs) [inline R/preprocessing.R:301]:
```suggestion
```

**Review by seabbs — DISMISSED** [review]

File: R/preprocessing.R [hunk 2]
```diff
@@ -177,3 +177,154 @@ default_fill_missing_obs <- function(data, obs, obs_column) {
...
+
+##' Filter leading zeros from a data set.
+##'
+##' @inheritParams estimate_infections
+##' @inheritParams fill_missing
+##' @return A data.table with leading zeros removed.
+##' @export
+##' @importFrom data.table setDT
+##' @examples
+##' cases <- data.table(
```
Comment (seabbs) [inline R/preprocessing.R:264]:
```suggestion
##' cases <- data.frame(
```

as `data.table` isn't loaded this currently errors

File: R/preprocessing.R (hunk 1 above)
Comment (sbfnk) [reply R/preprocessing.R:301]: arg sorry I thought I'd stopped emacs from doing that but clearly I haven't

**Review by seabbs — APPROVED** [review]

**Review by seabbs — DISMISSED** [review]

**Comment by seabbs** [comment]
I think there is an issue with the message formatting in the depreciation note.

**Comment by jamesmbaazam** [comment]
Might be a task for a separate issue but maybe we should showcase these functions in the examples of the exported functions.

File: NEWS.md [hunk 3]
```diff
@@ -19,6 +19,10 @@
 - A bug was fixed where an internal function for applying a default cdf cutoff failed due to a difference a vector length issue. By @jamesmbaazam in #858 and reviewed by @sbfnk.
 - All parameters have been changed to the new parameter interface. By @sbfnk in #871 and reviewed by @seabbs.
 
+## Package changes
+
+- The internal functions `create_clean_reported_cases()` has been broken up into several functions, with relevant ones `filter_leading_zeros()`, `add_breakpoints()` and `apply_zero_threshold()` exposed to the user. By @sbfnk in #884 and reviewed by @seabbs.
```
Comment (jamesmbaazam) [inline NEWS.md:24]:
```suggestion
- The internal functions `create_clean_reported_cases()` has been broken up into several functions, with relevant ones `filter_leading_zeros()`, `add_breakpoints()` and `apply_zero_threshold()` exposed to the user. By @sbfnk in #884 and reviewed by @seabbs and @jamesmbaazam.
```

File: R/preprocessing.R [hunk 4]
```diff
@@ -177,3 +177,154 @@ default_fill_missing_obs <- function(data, obs, obs_column) {
   }
   return(data)
 }
+
+##' Add missing values for future dates
+##'
+##' @param accumulate The number of days to accumulate when generating posterior
+##'   prediction, e.g. 7 for weekly accumulated forecasts.
+##' @inheritParams add_horizon
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/preprocessing.R:185]: ah yes, need to change to `fill_missing`

**Comment by sbfnk** [comment]
Passing tests now so just needs someone to approve to enter the merge queue @seabbs @jamesmbaazam

**Review by jamesmbaazam — APPROVED** [review]

**Pull request merged**